Fix #2103 and #2110 by clearly defining motor state and how it's modified

knandersen opened the pull request:
Resolves #2103: Motor state handling can be improved
Resolves #2110 : Changing motor speed/direction after running for timed motor commands makes it run forever

knandersen committed:
Major change of motor state handling to increase reliability, clear responsibility of handling state, and readability of the code.

BoostMotor now has a status getter/setter that replaces isOn() and is responsible for clearing various motor state parameters.

A new BoostMotorState-enum contains the possible states a motor can be in.

Since time-based motor commands really just trigger a BoostMotor.turnOn(), it's the opcodes that are responsible for setting the motor state.

knandersen committed:
Corrected documentation for BoostMotor._clearRotationState()

knandersen committed:
removed power wrongly being set in setMotorDirection()

knandersen committed:
Added state-change for Boost.stopAllMotors()

knandersen commented:
Two things I'm wondering:

1. Ideally the motor state should be set in the `BoostMotor`-functions rather than in the opcodes, since the former are "closer" to the actual BLE-commands getting sent. This, however, is hard to do, because of the interaction between `turnOnFor()` and `turnOn()`, since the latter doesn't know that it's called by the former, and thus can't know what motor-state to change to.

2. Are there actually scenarios where we're interested in retaining motor-state relating to either timed- or rotation-based behavior? With this PR, `BoostMotor.status = value` has a `switch/case` that determines whether to clear motor state relating to `time` or `rotation`, but perhaps this is unnecessary, and there should just be a `_clearMotorState()`-function that resets both, since a motor can only be doing one thing at a time.

Thoughts? cc @ericrosenbaum @evhan55

ericrosenbaum commented:
Great questions! From our discussion:

1. I like the idea of the BoostMotor object managing the motor state. One solution for the issue with `turnOnFor()` would be to make a new function like `turnOnForever`, and call `turnOn` internally from the motor object only (i.e. make `turnOn` "private"). Then, `turnOnFor` and `turnOnForever` would set the motor state, and both would call `turnOn` (which would not).

2. Yeah! seems like a good idea have a function that clears the motor state.

knandersen committed:
This commit addresses point 1 from the discussion with @ericrosenbaum around moving the setting of motor-state into the BoostMotor-class rather than having it in the opcodes.

- turnOn() renamed to _turnOn() and marked as a private function, i.e. it should only be called by BoostMotor-functions, not opcodes.
- New function turnOnForever() to be called by opcodes.
- turnOff() now sets the motor state.
- _clearRotationState now does a check for null rather than a truthy value
- all motor state setting removed from Boost-class and opcodes: stopAllMotors(), motorOnFor(), motorOnForRotation(), motorOn(), motorOff()
- turnOnForever(), turnOnFor() and turnOnForDegrees() now have a resetState-parameter with the default value of true. This allows the setMotorPower() and setMotorDirection()-functions to not reset state, to avoid them resolving the promises of the original motor commands that they are affecting.

knandersen committed:
changes BoostMotor.status(value) to reset all motor state

knandersen committed:
Added a special case in motorOnForRotation() to avoid hanging blocks if power is 0

knandersen committed:
Simplified the return value for when power is 0 in motorOnForRotation()

evhan55 commented:
Tested this branch and:

- Fixes the issue in #2110 for me!  I was able to reproduce it using the STR stack from #2110 without this change, and then I saw it fixed with this code.
- The new motor state handling feels very readable to me.  I only have one question from reading through it a couple of times, which I commented in the code directly.

Looks good to me!

evhan55 commented on src/extensions/scratch3_boost/index.js:
```diff
@@ -439,10 +457,11 @@ class BoostMotor {
-    turnOn () {
-        if (this._power === 0) return;
+    _turnOn () {
+        if (this.power === 0) return;
```
`this.power === 0;` is distinct from `BoostMotorState.OFF` ?

knandersen replied:
Yeah, `this.power` describes the motor's power between 0-100, where as `BoostMotorState.OFF` is an enumerated value.

evhan55 commented on src/extensions/scratch3_boost/index.js:
```diff
@@ -1729,15 +1763,18 @@ class Scratch3BoostBlocks {
-                    } else {
-                        motor.turnOn();
-                    }
+                switch (motor.status) {
```
This refactor feels really good!

evhan55 reviewed (approved):
Other than wondering how `this.power === 0` is different from `BoostMotorState.OFF`, everything lgtm!
